## 1

Acessar o link abaixo e realizar o *download* do arquivo **"datasets_diabetes.csv"**

https://drive.google.com/drive/folders/1-KhRl5Q1QK8d_N4AMRRoZti8gdQFGLpX?usp=sharing

## 2

Para a implementação dos algoritmos, utilize as definições abaixo:
<br><br>
**Algoritmo KNN:**

```clf_KNN = KNeighborsClassifier(n_neighbors=5)```
<br><br>

**Algoritmo Árvore de Decisão:**

```clf_arvore = DecisionTreeClassifier(random_state=1)```
<br><br>

**Algoritmo Floresta Randômica:**

```clf_floresta = RandomForestClassifier(max_depth=10, random_state=1)```
<br><br>

**Algoritmo SVM:**

```clf_svm=SVC(gamma='auto',random_state=1)```
<br><br>

**Algoritmo Rede MLP:**

```clf_mlp = MLPClassifier(solver='lbfgs', alpha=1e-5, hidden_layer_sizes=(5, 5), random_state=1)```

**Observações:**

- Utilize como entrada as colunas: **Pregnancies, Glucose, BloodPressure, SkinThickness, Insulin, BMI, DiabetesPedigreeFunction e Age**. 
A saída para os algoritmos deve ser a coluna **Outcome**.
<br><br>

- Realize a carga dos valores utilizando o método ```pandas.read_csv()```
<br><br>

- Utilize, para normalização dos dados, as definições: 
    - ```normaliza = MinMaxScaler() #objeto para a normalização```
    - ```entradas_normalizadas=normaliza.fit_transform(entradas)```
<br><br>

- Utilize, para divisão entre treinamento e teste do algoritmo, as definições: ```train_test_split(entradas_normalizadas, saida, test_size=0.30,random_state=42)```
<br><br>

- Utilize esta sequência de operações para chegar no resultado final:
    - Divida os dados entre entrada e saída; 
    - Normalize apenas as entradas utilizando o MinMaxScaler;
    - Aplique a divisão entre o treinamento e o teste com o train_test_split.
<br><br>

- Utilize os dados de “teste” para avaliar as previsões de classificação dos modelos.

In [3]:
import pandas as pd

In [4]:
df = pd.read_csv("datasets_diabetes.csv")

In [9]:
df.head()

Pregnancies  Glucose  BloodPressure  ...  DiabetesPedigreeFunction  Age  Outcome
0            6      148             72  ...                     0.627   50        1
1            1       85             66  ...                     0.351   31        0
2            8      183             64  ...                     0.672   32        1
3            1       89             66  ...                     0.167   21        0
4            0      137             40  ...                     2.288   33        1

[5 rows x 9 columns]

## Q1
Após a utilização da biblioteca pandas para a leitura dos dados e o método pandas.info() sobre o *dataframe* construído é CORRETO afirmar que:

In [15]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 768 entries, 0 to 767
Data columns (total 9 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   Pregnancies               768 non-null    int64  
 1   Glucose                   768 non-null    int64  
 2   BloodPressure             768 non-null    int64  
 3   SkinThickness             768 non-null    int64  
 4   Insulin                   768 non-null    int64  
 5   BMI                       768 non-null    float64
 6   DiabetesPedigreeFunction  768 non-null    float64
 7   Age                       768 non-null    int64  
 8   Outcome                   768 non-null    int64  
dtypes: float64(2), int64(7)
memory usage: 54.1 KB


Resposta:

Não foram encontrados valores nulos após a leitura dos dados

## Q2
Quantas linhas e colunas, respectivamente, existem no dataset?

In [6]:
df.shape

(768, 9)

Resposta:

768 linhas e 9 colunas

## Q3

Quantas colunas do tipo "float64" existem no dataset após a carga dos dados através do método ```pandas.read_csv()```?

Resposta:

2

## Q4

Qual o valor médio presente na coluna "Age"?

In [8]:
df["Age"].mean()

33.240885416666664

## Q5

Utilizando as entradas e a saída como apresentado no enunciado do trabalho, quantos valores da variável de saída "1" existem no conjunto de dados?

In [10]:
df["Outcome"].sum()

268

## Q6

Utilizando o método “train_test_split” como apresentado no enunciado, quantas instâncias (linhas) foram selecionadas para o conjunto de dados de treinamento do modelo?

In [25]:
from sklearn import neighbors, datasets
import pandas as pd
import numpy as np

In [26]:
# transformação dos dados em array
X = df.iloc[:, :-1].values # dados de entrada
y = df.iloc[:, 8].values # outcome

In [37]:
# normalização
from sklearn.preprocessing import MinMaxScaler

normaliza = MinMaxScaler() #objeto para a normalização
entradas_normalizadas=normaliza.fit_transform(X_train)

In [38]:
# realiza a divisão dos dados entre treinamento e teste
from sklearn.model_selection import train_test_split # função que realiza a divisão do dataset
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.30, random_state=42) # divide 30% para teste

In [29]:
# divisão dos dados entre treino e teste
# train_test_split(entradas_normalizadas, saida, test_size=0.30,random_state=42)

In [39]:
np.shape(X_train)

(537, 8)

## Q7

Após dividir as colunas do dataframe entre entrada ee saída, aplicar a normalização dos dados como apresentado no enunciado (MinMaxScaler()) e dividir esses dados entre treinamento e teste, aplique o algoritmo **KNN**. Qual é, aproximadamente, a acurácia do modelo?

In [40]:
# treina o modelo
from sklearn.neighbors import KNeighborsClassifier
clf_KNN = KNeighborsClassifier(n_neighbors=5)
clf_KNN.fit(X_train, y_train)

KNeighborsClassifier(algorithm='auto', leaf_size=30, metric='minkowski',
                     metric_params=None, n_jobs=None, n_neighbors=5, p=2,
                     weights='uniform')

In [41]:
# previsão
y_pred = clf_KNN.predict(X_test)

In [42]:
# matriz de confusão para comparar o modelo criado
from sklearn.metrics import classification_report, confusion_matrix
print(confusion_matrix(y_test, y_pred))
print(classification_report(y_test, y_pred))

[[114  37]
 [ 35  45]]
              precision    recall  f1-score   support

           0       0.77      0.75      0.76       151
           1       0.55      0.56      0.56        80

    accuracy                           0.69       231
   macro avg       0.66      0.66      0.66       231
weighted avg       0.69      0.69      0.69       231



## Q8

Após dividir as colunas do dataframe entre entrada e saída, aplicar a normalização dos dados como apresentado no enunciado (MinMaxScaler()) e dividir esses dados entre treinamento e teste, aplique o algoritmo Árvore de Decisão. Qual é, aproximadamente, a acurácia do modelo?

In [34]:
# treinamento
from sklearn.tree import DecisionTreeClassifier
clf_arvore = DecisionTreeClassifier(random_state=1)
clf_arvore.fit(X_train, y_train)

DecisionTreeClassifier(ccp_alpha=0.0, class_weight=None, criterion='gini',
                       max_depth=None, max_features=None, max_leaf_nodes=None,
                       min_impurity_decrease=0.0, min_impurity_split=None,
                       min_samples_leaf=1, min_samples_split=2,
                       min_weight_fraction_leaf=0.0, presort='deprecated',
                       random_state=1, splitter='best')

In [35]:
y_pred = clf_arvore.predict(X_test)

In [36]:
print(confusion_matrix(y_test, y_pred))
print(classification_report(y_test, y_pred))

[[105  46]
 [ 25  55]]
              precision    recall  f1-score   support

           0       0.81      0.70      0.75       151
           1       0.54      0.69      0.61        80

    accuracy                           0.69       231
   macro avg       0.68      0.69      0.68       231
weighted avg       0.72      0.69      0.70       231



## Q9

Após dividir as colunas do dataframe entre entrada e saída, aplicar a normalização dos dados como apresentado no enunciado (MinMaxScaler()) e dividir esses dados entre treinamento e teste, aplique o algoritmo Floresta Randômica. Qual é, aproximadamentee, a acurácia do modelo?

In [ ]:
from sklearn import RandomForestClassifier
clf_floresta = RandomForestClassifier(max_depth=10, random_state=1)

## Q10

Após dividir as colunas do dataframe entre entrada e saída, aplicar a normalização dos dados como apresentado no enunciado (MinMaxScaler()) e dividir esses dados entre treinamento e teste, aplique o algoritmo SVM. Qual é, aproximadamentee, a acurácia do modelo?

## Q11

Após dividir as colunas do dataframe entre entrada e saída, aplicar a normalização dos dados como apresentado no enunciado (MinMaxScaler()) e dividir esses dados entre treinamento e teste, aplique o algoritmo MLP. Qual é, aproximadamentee, a acurácia do modelo?

## Q12

Dentre todos oos modelos implementadoos seguindo o enunciado deste trabalho prático, qual foi aquele que alcançou maior acurácia?

## Q13

Qual dos modelos implementados neste trabalho prático é classificado como um modelo de aprendizado de máquina não supervisionado

Resposta: 

Nenhum deles


KNN = supervisionado
SVM = supervisionado

## Q14

Implemente o código 1 presente no enunciado do trabalho.
Sobre esse cóodiggo, é INCORRETO afirmar que:



In [ ]:
# CÓDIGO 1

import threading
import time
from random import randint

def funcao_1(num):
    n=num
    while n>0:
        n-=1
        print("n_1: {}".format(n))
        time.sleep(randint(0,2))

def funcao_2(num):
    n=num
    while n<100:
        n+=1
        print("n_2: {}".format(n))
        time.sleep(randint(0,2))

if __name__ == "__main__":
    t1 = threading.Thread(target=funcao_1, args=(100,))
    t2 = threading.Thread(target=funcao_2, args=(0,))

    t1.start()
    t2.start()

    t1.join()
    t2.join()

    print("Fim!")

## Q15

Considere o código 1. Inverta os parâmetros de entrada para as funções 1 e 2 (entrada funcao_1=0 e entrada funcao_2=100). Após essas modificações, é INCORRETO afirmar que:
